ethash/src/shared.rs

```diff
@@ -0,0 +1,85 @@
+use std::mem;
+use primal::is_prime;
+
+pub const DATASET_BYTES_INIT: u64 = 1 << 30;
+pub const DATASET_BYTES_GROWTH: u64 = 1 << 23;
+pub const CACHE_BYTES_INIT: u64 = 1 << 24;
+pub const CACHE_BYTES_GROWTH: u64 = 1 << 17;
+
+pub const ETHASH_EPOCH_LENGTH: u64 = 30000;
+pub const ETHASH_CACHE_ROUNDS: usize = 3;
+pub const ETHASH_MIX_BYTES: usize = 128;
+pub const ETHASH_ACCESSES: usize = 64;
+pub const ETHASH_DATASET_PARENTS: u32 = 256;
+pub const NODE_DWORDS: usize = NODE_WORDS / 2;
+pub const NODE_WORDS: usize = NODE_BYTES / 4;
+pub const NODE_BYTES: usize = 64;
+
+pub fn epoch(block_number: u64) -> u64 {
+	block_number / ETHASH_EPOCH_LENGTH
+}
+
+static CHARS: &'static [u8] = b"0123456789abcdef";
+pub fn to_hex(bytes: &[u8]) -> String {
+	let mut v = Vec::with_capacity(bytes.len() * 2);
+	for &byte in bytes.iter() {
+		v.push(CHARS[(byte >> 4) as usize]);
+		v.push(CHARS[(byte & 0xf) as usize]);
+	}
+
+	unsafe { String::from_utf8_unchecked(v) }
+}
+
+pub fn get_cache_size(block_number: u64) -> usize {
+	// TODO: Memoise
+	let mut sz: u64 = CACHE_BYTES_INIT + CACHE_BYTES_GROWTH * (block_number / ETHASH_EPOCH_LENGTH);
+	sz = sz - NODE_BYTES as u64;
+	while !is_prime(sz / NODE_BYTES as u64) {
+		sz = sz - 2 * NODE_BYTES as u64;
+	}
+	sz as usize
+}
+
+pub fn get_data_size(block_number: u64) -> usize {
+	// TODO: Memoise
+	let mut sz: u64 = DATASET_BYTES_INIT + DATASET_BYTES_GROWTH * (block_number / ETHASH_EPOCH_LENGTH);
+	sz = sz - ETHASH_MIX_BYTES as u64;
+	while !is_prime(sz / ETHASH_MIX_BYTES as u64) {
+		sz = sz - 2 * ETHASH_MIX_BYTES as u64;
+	}
+	sz as usize
+}
+
+pub struct Node {
+	pub bytes: [u8; NODE_BYTES],
+}
+
+impl Default for Node {
+	fn default() -> Self {
+		Node { bytes: [0u8; NODE_BYTES] }
+	}
+}
+
+impl Clone for Node {
+	fn clone(&self) -> Self {
+		Node { bytes: *&self.bytes }
+	}
+}
+
+impl Node {
+	pub fn as_words(&self) -> &[u32; NODE_WORDS] {
```

Comment:
Yeah, the way I'd go is that if safe code can not cause unsafety, no matter what data it passes to the function, then then function is safe. Otherwise it's `unsafe`. Whether or not the code inside uses `unsafe` is immaterial. `Vec::set_len` is unsafe despite being 100% "safe" code. As for the alignment issue, I thought that one-field structs were guaranteed to have the same representation as their inner type. The assertion should be `mem::size_of::<[u8; NODE_BYTES]>() == mem::size_of::<[u64; NODE_DWORDS]/[u32; NODE_WORDS]>()`, I guess that's not handled by the transmute since it's transmuting pointers. I should note that if this is a problem then it's a problem in the current version of the software, this is no worse than it (unlike the file reading issue, which is a regression), but that's not a reason not to change it.